packages/api/src/upload.js

```diff
@@ -0,0 +1,92 @@
+/* eslint-env serviceworker */
+import { gql } from '@web3-storage/db'
+import { JSONResponse } from './utils/json-response.js'
+import { toFormData } from './utils/form-data.js'
+import { LOCAL_ADD_THRESHOLD } from './constants.js'
+import { toPinStatusEnum } from './utils/pin.js'
+
+const IMPORT_CAR = gql`
+  mutation importCar($data: ImportCarInput!) {
+    importCar(data: $data) {
+      content {
+        _id
+      }
+    }
+  }
+`
+
+/**
+ * Post a File/Directory.
+ *
+ * @param {import('./user').AuthenticatedRequest} request
+ * @param {import('./env').Env} env
+ */
+export async function uploadPost (request, env) {
+  const { user, authToken } = request.auth
+  const { headers } = request
+  const contentType = headers.get('content-type') || ''
+
+  let cid
+  let dagSize
+  let name = headers.get('x-name')
+  if (!name || typeof name !== 'string') {
+    name = `Upload at ${new Date().toISOString()}`
+  }
+
+  if (contentType.includes('multipart/form-data')) {
+    const form = await toFormData(request)
+    const files = /** @type {File[]} */ (form.getAll('file'))
+    const dirSize = files.reduce((total, f) => total + f.size, 0)
+
+    const dir = await env.cluster.addDirectory(files, {
+      metadata: { size: dirSize.toString() }
+    })
+    const { cid: cCId, size } = dir[dir.length - 1]
+
+    cid = cCId
+    dagSize = size
+  } else {
+    const blob = await request.blob()
+    if (blob.size === 0) {
+      throw new Error('Empty payload')
+    }
+
+    const { cid: cCId, size } = await env.cluster.add(blob, {
+      metadata: { size: blob.size.toString() },
+      // When >2.5MB, use local add, because waiting for blocks to be sent to
+      // other cluster nodes can take a long time. Replication to other nodes
+      // will be done async by bitswap instead.
+      local: blob.size > LOCAL_ADD_THRESHOLD
+    })
+
+    cid = cCId
+    dagSize = size
```

Comment:
```suggestion
    const entry = await env.cluster.add(blob, {
      metadata: { size: blob.size.toString() },
      // When >2.5MB, use local add, because waiting for blocks to be sent to
      // other cluster nodes can take a long time. Replication to other nodes
      // will be done async by bitswap instead.
      local: blob.size > LOCAL_ADD_THRESHOLD
    })

    cid = entry.cid
    dagSize = entry.size
```